### PR TITLE
Adding route for FAQ

### PR DESCRIPTION
**Pull request by Kriti-bit**
## Issue that this pull request solves

 Closes: #501 

## Proposed changes

I have added the FAQ route, however I need to structure it a little more. If there is anything else that needs to be changed, please let me know.

_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update (Documentation content changed)
- [ ] Other (please describe): 

## Checklist

_Put an `x` in the boxes that apply_

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] My changes does not break the current system and it passes all the current test cases.

### CODE REVIEW

**Review by Kajol-Kumari — CHANGES_REQUESTED** [review]
Hey @Kriti-bit please remove the `node_modules` commit from this PR. You can use `git checkout HEAD -- ./node_modules` For details, please checkout [here](https://stackoverflow.com/questions/7147270/hard-reset-of-a-single-file)

File: backend/app/models/faq.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        type: Boolean,
+    },
+    category:
+    {
+        type:String
+    }
+})
+
+module.exports = mongoose.model('faq', faqSchema)
```
Comment (jackfrost13) [inline backend/app/models/faq.js:30]: db collection name should be `FAQ`

File: backend/app/models/faq.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        type: String,
+        trim:true,
+        required:true,
+        unique:true
+    },
+    isActive:
+    {
+        type: Boolean,
+    },
+    category:
```
Comment (jackfrost13) [inline backend/app/models/faq.js:24]: Please change it to `tags` which will be an array of strings

File: backend/app/routes/faq/index.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+const {authMiddleware, generateJWT} = require('../../../helpers/middlewares/auth');
+const express = require ("express");
+const FAQ = require('../../models/faq');
+const router = express.Router();
+
+
+router.use('/', authMiddleware , async (req,res,next) => {
```
Comment (jackfrost13) [inline backend/app/routes/faq/index.js:7]: You need to add the Validation schema, check other routes for it
Also extract the function to a seperate file

File: backend/app/routes/faq/index.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    try {        
+        const newFAQ = new FAQ({
```
Comment (jackfrost13) [inline backend/app/routes/faq/index.js:9]: you can directly use the req.body and need not create a new obj like this or use const FAQ = req.body
that should do the work

File: backend/app/routes/faq/index.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+            isActive: req.body.isActive,
+            category: req.body.category,
+        })
+        
+        const faq = await newFAQ.save()
+        res.json(faq);
+        res.status(200);
+        return next();
+    } catch (err) {
+      console.error(err.message)
```
Comment (jackfrost13) [inline backend/app/routes/faq/index.js:21]: You should use our error handler

File: backend/app/routes/faq/index.js (hunk 3 above)
Comment (jackfrost13) [reply backend/app/routes/faq/index.js:9]: This is pending

File: backend/app/routes/faq/index.js (hunk 4 above)
Comment (jackfrost13) [reply backend/app/routes/faq/index.js:11]: Also this

**Review by jackfrost13 — CHANGES_REQUESTED** [review]
Please remove node modules from your commit

**Comment by Kriti-bit** [comment]
I'm still working on it, I'll let you know once all the changes are done

**Comment by jackfrost13** [comment]
> I'm still working on it, I'll let you know once all the changes are done

Oh got it. Please request for a review when you are done

**Comment by Kriti-bit** [comment]
@jackfrost13 are there any other changes that required?

File: backend/app/routes/faq/@validationSchema/index.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+const Joi = require('joi');
+
+const faqSchema = Joi.object().keys({
+  question : Joi.string().required(),
+  answer : Joi.string().required(),
+});
```
Comment (jackfrost13) [inline backend/app/routes/faq/@validationSchema/index.js:6]: tags can be also present

File: backend/app/routes/faq/@validationSchema/index.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+const Joi = require('joi');
+
+const faqSchema = Joi.object().keys({
```
Comment (jackfrost13) [inline backend/app/routes/faq/@validationSchema/index.js:3]: this should be FAQValidationSchema

File: backend/app/routes/faq/faq.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
+const {authMiddleware, generateJWT} = require('../../../helpers/middlewares/auth');
+const express = require ("express");
+const FAQ = require('../../models/faq');
+const router = express.Router();
+const { ErrorHandler } = require('../../../helpers/error');
+const constants = require('../../../constants');
+
```
Comment (jackfrost13) [inline backend/app/routes/faq/faq.js:7]: file name should be post.js

File: backend/app/routes/faq/faq.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
+const {authMiddleware, generateJWT} = require('../../../helpers/middlewares/auth');
```
Comment (jackfrost13) [inline backend/app/routes/faq/faq.js:1]: remove unwanted imports

File: backend/app/routes/faq/faq.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        if(tags) faq.tags = tags; 
+
+        const Faq = new FAQ(faq)
+
+        await Faq.save();
+       
+        res.json(Faq);
+        res.status(200);
+        return next();
+    } catch (err) {
```
Comment (jackfrost13) [inline backend/app/routes/faq/faq.js:31]: avoid using try catch, instead use await-to-js for async flows

File: backend/app/routes/faq/faq.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+const router = express.Router();
+const { ErrorHandler } = require('../../../helpers/error');
+const constants = require('../../../constants');
+
+module.exports = async (req, res, next) => {
+    try {        
+        const {
+            question,
+            answer,
+            isActive,
```
Comment (jackfrost13) [inline backend/app/routes/faq/faq.js:13]: you are trying to set fields which are not present in the request body, check that

File: backend/app/routes/faq/faq.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+            tags,
+            }=req.body;
+
+        const faq = {};
+
+        faq.question = question;
+        faq.answer = answer;
+        if(isActive) faq.isActive = isActive;
+        if(tags) faq.tags = tags;
```
Comment (jackfrost13) [inline backend/app/routes/faq/faq.js:22]: These can be done in a single assignment, no need of first consuming them in variables and then in json

File: backend/app/routes/faq/faq.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+        const Faq = new FAQ(faq)
+
+        await Faq.save();
+       
+        res.json(Faq);
+        res.status(200);
+        return next();
+    } catch (err) {
+        const error = new ErrorHandler(constants.ERRORS.UNEXPECTED, {
```
Comment (jackfrost13) [inline backend/app/routes/faq/faq.js:32]: Error should be DB error

**Review by jackfrost13 — CHANGES_REQUESTED** [review]

**Comment by jackfrost13** [comment]
@Kriti-bit Please add testing file also(mocha-chai) along with required documentation.
Request for review when done.
Also you need to improve on your commit messages

**Comment by Kriti-bit** [comment]
@jackfrost13 are there any other changes that required?

File: backend/app/models/faq.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+  answer: {
+    type: String,
+    trim: true,
+    required: true,
+  },
+  isActive: {
+    type: Boolean,
+  },
+  tags: {
+    type: Array,
```
Comment (jackfrost13) [inline backend/app/models/faq.js:21]: was [String] not working?

File: backend/app/routes/faq/@validationSchema/index.js [hunk 15]
```diff
@@ -0,0 +1,9 @@
+const Joi = require('joi');
+
+const FAQValidationSchema = Joi.object().keys({
+  question: Joi.string().required(),
+  answer: Joi.string().required(),
+  tags: Joi.string().required(),
```
Comment (jackfrost13) [inline backend/app/routes/faq/@validationSchema/index.js:6]: tags is an array

File: backend/app/routes/faq/index.js [hunk 16]
```diff
@@ -0,0 +1,8 @@
+const router = require('express').Router({ mergeParams: true });
+const faq = require('./post');
+const FAQValidationSchema = require('./@validationSchema');
+const validation = require('../../../helpers/middlewares/validation');
+
+router.use('/faq', validation(FAQValidationSchema), faq);
```
Comment (jackfrost13) [inline backend/app/routes/faq/index.js:6]: this shouldn't be router.use

File: backend/app/routes/faq/post.js [hunk 17]
```diff
@@ -0,0 +1,20 @@
+const to = require('await-to-js').default;
+const FAQ = require('../../models/faq');
+const { ErrorHandler } = require('../../../helpers/error');
+const constants = require('../../../constants');
+
+module.exports = async (req, res, next) => {
+  const [err] = await to(FAQ.create({ ...req.body }));
```
Comment (jackfrost13) [inline backend/app/routes/faq/post.js:7]: you are not taking isActive in req body, take care of that, you need to assign it here i guess

**Review by jackfrost13 — CHANGES_REQUESTED** [review]

File: backend/app/models/faq.js (hunk 14 above)
Comment (Kriti-bit) [reply backend/app/models/faq.js:21]: Arrays and [String] both of them are working

File: backend/app/routes/faq/post.js (hunk 17 above)
Comment (Kriti-bit) [reply backend/app/routes/faq/post.js:7]: I tested it, it is taking isActive as well

Comment (jackfrost13) [reply backend/app/routes/faq/post.js:7]: I mean, in the req body you will not get isActive

Comment (Kriti-bit) [reply backend/app/routes/faq/post.js:7]: I'm not sure if I understood you correctly, the screenshots below are of post.js and the response, that I mentioned earlier.

image

image

Comment (jackfrost13) [reply backend/app/routes/faq/post.js:7]: Yes, I understood
You will not get isActive field in the request body always

Comment (jackfrost13) [reply backend/app/routes/faq/post.js:7]: how will it work when request body contains
question answer and tags only

Comment (Kriti-bit) [reply backend/app/routes/faq/post.js:7]: Do you mean something like this?
image

**Comment by jackfrost13** [comment]
yes

File: backend/app/routes/faq/post.js (hunk 17 above)
Comment (Kriti-bit) [reply backend/app/routes/faq/post.js:7]: I've set it to true by default in case it is not provided in req.body. Would that be alright?

**Comment by Kriti-bit** [comment]
@jackfrost13 are there any other changes that are required?

**Comment by Kriti-bit** [comment]
Hi @jackfrost13, any other change required for this PR?

**Review by jackfrost13 — APPROVED** [review]

**Pull request merged**